File: src/main/java/com/zt/kafka/basic/string/Producer.java
package com.zt.kafka.basic.string;

import com.zt.kafka.basic.KafkaProperties;
import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.ByteArraySerializer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class Producer extends Thread {
    private final KafkaProducer producer;
    private final String topic;
    private final Boolean isAsync;

    public Producer(String topic, Boolean isAsync) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaProperties.BOOTSTRAPS);
        props.put(ProducerConfig.CLIENT_ID_CONFIG, "DemoProducer");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
        producer = new KafkaProducer<String, byte[]>(props);
        this.topic = topic;
        this.isAsync = isAsync;
    }

    public void run() {
        int messageNo = 1;
        while (messageNo <= 10) {

            String messageStr = null;

            byte[] bytes = new byte[10];
                for (int i = 0; i <10 ; i++) {
                    bytes[i] = 'a';
            }
            messageStr = new String(bytes);
            long startTime = System.currentTimeMillis();
            if (isAsync) { // Send asynchronously
                producer.send(new ProducerRecord<String, byte[]>(topic,
                        bytes), new DemoCallBack(startTime, messageNo, messageStr));
            } else { // Send synchronously
                try {
                    Future<RecordMetadata> future = producer.send(
                            new ProducerRecord(
                                    topic,
                                    bytes));
                    RecordMetadata recordMetadata = future.get();
                    System.out.println("Sent message: (" + messageNo + ", " + messageStr + ") to partition: "
                            + recordMetadata.partition() + " , offset: " + recordMetadata.offset()   );
                } catch (InterruptedException | ExecutionException e) {
                    e.printStackTrace();
                }
            }
            ++messageNo;


            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public static void main(String[] args) throws Exception{

        String topic = "KafkaLowApiStream3";

        Producer producerThread = new Producer(topic, false);
        producerThread.start();
//
//        Thread.sleep(100);

//        Consumer consumerThread1 = new Consumer("group-7", topic);
//        consumerThread1.start();

    }
}

class DemoCallBack implements Callback {

    private final long startTime;
    private final int key;
    private final String message;

    public DemoCallBack(long startTime, int key, String message) {
        this.startTime = startTime;
        this.key = key;
        this.message = message;
    }

    /**
     * A callback method the user can implement to provide asynchronous handling of request completion. This method will
     * be called when the record sent to the server has been acknowledged. Exactly one of the arguments will be
     * non-null.
     *
     * @param metadata  The metadata for the record that was sent (i.e. the partition and offset). Null if an error
     *                  occurred.
     * @param exception The exception thrown during processing of this record. Null if no error occurred.
     */
    public void onCompletion(RecordMetadata metadata, Exception exception) {
        long elapsedTime = System.currentTimeMillis() - startTime;
        if (metadata != null) {
            System.out.println(
                    "message(" + key + ", " + message + ") sent to partition(" + metadata.partition() +
                            "), " +
                            "offset(" + metadata.offset() + ") in " + elapsedTime + " ms");
        } else {
            exception.printStackTrace();
        }
    }
}

File: component/practise/src/main/scala/com/zt/scala/kafka/string/Producer.scala
package com.zt.scala.kafka.string

import java.util.Properties
import java.util.concurrent.Future

import com.zt.scala.constant.KafkaProperties
import org.apache.kafka.clients.producer._
import org.apache.kafka.common.serialization.StringSerializer

class Producer(val pTopic: String, pIsAsync: Boolean) extends Thread {

  private var kafkaProducer: KafkaProducer[String, String] = _
  private var topic: String = pTopic
  private var isAsync: Boolean = pIsAsync

  val props = new Properties()
  props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaProperties.BOOTSTRAPS)
  props.put(ProducerConfig.CLIENT_ID_CONFIG, "DemoProducer")
  props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  val producer = new KafkaProducer[String, String](props)

  override def run(): Unit = {
    var messageNo: Int = 1
    while (messageNo <= 100) {
      val messageStr = String.valueOf(System.currentTimeMillis())

      val bytes = new Array[Byte](1024*1000)
      for (i <- 0 to bytes.length - 1) {
        bytes(i) = 'a'
      }

      println(topic + " " + isAsync)

      val startTime = System.currentTimeMillis()
      if (isAsync) {
        producer.send(new ProducerRecord[String, String](
          topic,
          new String(bytes)), new MyCallBack(startTime, messageNo, messageStr))
      } else {

        try {
          val future: Future[RecordMetadata] = producer.send(
            new ProducerRecord(topic, messageStr))
          val recordMetadata = future.get()
          System.out.println("Sent message: (" + messageNo + ", " + messageStr + ") to partition: "
            + recordMetadata.partition() + " , offset: " + recordMetadata.offset());
        } catch {
          case e: Throwable => throw e
        }
      }
      messageNo += 1
    }
  }


}

class MyCallBack(sTime: Long, k: Int, msg: String) extends Callback {

  private val startTime: Long = sTime
  private val key: Int = k
  private val message: String = msg


  override def onCompletion(metadata: RecordMetadata, e: Exception): Unit = {
    val elapsedTime = System.currentTimeMillis() - startTime
    if (metadata != null) {
      println("message(" + key + ", " + message + ") sent to partition(" + metadata.partition() +
        "), " +
        "offset(" + metadata.offset() + ") in " + elapsedTime + " ms")
    } else {
      e.printStackTrace()
    }
  }
}

object Te extends App {
  val producer = new Producer("test3", true)
  producer.start()
}